CONFIG.md
# Configuration
The first time brooce runs, it will create a `~/.brooce` dir in your home directory with a default `~/.brooce/brooce.conf` config file. That default config file is shown here, and we will explain what it all means in the following sections.

```json
{
  "cluster_name": "brooce",
  "global_job_options": {
    "timeout": 3600,
    "maxtries": 1,
    "requeuedelayed": 60,
    "redislogexpireafter": 604800
  },
  "web": {
    "addr": ":8080",
    "certfile": "",
    "keyfile": "",
    "username": "admin",
    "password": "afwyczvk",
    "no_auth": false,
    "no_log": false,
    "disable": false
  },
  "file_output_log": {
    "enable": false
  },
  "redis": {
    "host": "localhost:6379",
    "socket": "",
    "password": "",
    "db": 0
  },
  "suicide": {
    "enable": false,
    "command": "",
    "time": 0
  },
  "queues": [
    {
      "name": "common",
      "workers": 1,
      "job_options": {}
    }
  ],
  "basepath": "",
  "path": ""
}
```

### `cluster_name`
Leave this alone unless you want multiple sets of workers to share one redis server. Multiple brooce workers on separate machines can normally draw jobs from the same queue, but putting them in separate clusters will make them unaware of each other.

### `global_job_options`
Job options specified here will apply globally, unless overwritten in the queue job_options hash, or individually in the job. [See the list of all job options in README.md](README.md#job-options).
 
### `web.addr`
Where the web server is hosted. Defaults on port 8080 on all IPs that it can bind to.
 
### `web.certfile` / `web.keyfile`
Specify your HTTPS certificate and private key files here. If you have multiple certificate files, concatenate them into one file. If these are left blank, the web server will run in HTTP mode.
 
### `web.username` / `web.password`
We generate random login credentials the first time you run brooce, but you can change them here.
 
### `web.no_auth`
To run the web server with no authentication, leave username/password (above) blank, and set this to true. This is not recommended if you're having the web server listen on an internet-connected IP.
 
### `web.no_log`
By default, web interface access is logged to `~/.brooce/web.log`. Set this to true to disable web access logging.
 
### `web.disable`
Set to true to disable the web server.
 
### `file_output_log.enable`
By default, job stdout/stderr is only logged to redis for review through the web interface. If you turn this on, the `~/.brooce` folder will get a logfile for every worker.

### `redis.host` / `redis.password`
The hostname and password to access your redis server. Defaults to localhost and no-password.

### `redis.socket`
If specified, connect through a unix sock file instead of a hostname. Example: `/var/run/redis/redis-server.sock`

### `redis.db`
The db which will be used by brooce on your redis server. Defaults to 0.

### `suicide.enable` / `suicide.command` / `suicide.time`
For example, if you enabled suicide and set command to `"sudo shutdown -h now"` and time to `600`, you could shutdown your server after there haven't been any jobs for some time. Useful for shutting down idle EC2 instances. Keep in mind that the brooce program will need to have proper permissions to execute the given command, without additional prompts for passwords.

### `queues`
Brooce is multithreaded, and can listen for commands on multiple queues. For example, you could do the following to run 5 threads on the common queue and 2 more threads on the rare queue.

You can set per-queue job options in the `job_options` hash, as shown below. Per-queue options override `global_job_options`, and individual jobs can override the per-queue settings. [See the list of all job options in README.md](README.md#job-options).

```json
{
  "queues": [
    {
      "name": "common",
      "workers": 5
    },
    {
      "name": "rare",
      "workers": 2,
      "job_options": {
        "timeout": 7200,
        "maxtries": 5,
        "killondelay": true
      }
    }
  ]
}
```

### `basepath`
If you're running brooce's web server behind a proxy and exposing it as a folder (e.g. yourdomain.com/brooce/), you can set this to the directory path (e.g. `/brooce`).

### `path`
Add a given string to the brooce worker's PATH for running commands. For example, if you specify `/home/mydir/bin`, then now you can run a job as `mytask` instead of `/home/mydir/bin/mytask`.
 


config/config.go
package config

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strings"

	"brooce/myip"
	"brooce/util"
)

var BrooceConfigDir = os.Getenv("BROOCE_CONFIGDIR")
var BrooceLogDir = os.Getenv("BROOCE_LOGDIR")

type ConfigType struct {
	ClusterName    string `json:"cluster_name"`
	ClusterLogName string `json:"-"`
	ProcName       string `json:"-"`

	GlobalJobOptions JobOptions `json:"global_job_options"`

	Web struct {
		Addr     string `json:"addr"`
		BasePath string `json:"basepath"`
		CertFile string `json:"certfile"`
		KeyFile  string `json:"keyfile"`
		Username string `json:"username"`
		Password string `json:"password"`
		NoAuth   bool   `json:"no_auth"`
		NoLog    bool   `json:"no_log"`
		Disable  bool   `json:"disable"`
	} `json:"web"`

	FileOutputLog struct {
		Enable bool `json:"enable"`
	} `json:"file_output_log"`

	Redis struct {
		Host     string `json:"host"`
		Socket   string `json:"socket"`
		Password string `json:"password"`
		DB       int    `json:"db"`
	} `json:"redis"`

	Suicide struct {
		Enable  bool   `json:"enable"`
		Command string `json:"command"`
		Time    int    `json:"time"`
	} `json:"suicide"`

	Queues []Queue `json:"queues"`

	Path string `json:"path"`
}

type Queue struct {
	Name       string `json:"name"`
	Workers    int    `json:"workers"`
	JobOptions `json:"job_options"`
}

var Config = ConfigType{}

func (c *ConfigType) CSRF() string {
	return util.Md5sum(c.Web.Username + ":" + c.Web.Password)
}

func (c *ConfigType) JobOptionsForQueue(queue string) (opts JobOptions) {
	for _, q := range c.Queues {
		if q.Name == queue {
			opts = q.JobOptions
			return
		}
	}

	return
}

func (q *Queue) DeepJobOptions() (j JobOptions) {
	j.Merge(q.JobOptions)
	j.Merge(Config.GlobalJobOptions)
	j.Merge(DefaultJobOptions)
	return
}

func (q *Queue) PendingList() string {
	return fmt.Sprintf("%s:queue:%s:pending", Config.ClusterName, q.Name)
}

func (q *Queue) DoneList() string {
	return fmt.Sprintf("%s:queue:%s:done", Config.ClusterName, q.Name)
}

func (q *Queue) FailedList() string {
	return fmt.Sprintf("%s:queue:%s:failed", Config.ClusterName, q.Name)
}

func (q *Queue) DelayedList() string {
	return fmt.Sprintf("%s:queue:%s:delayed", Config.ClusterName, q.Name)
}

// this use of init sucks, but we'll have to fix every "var redisClient = myredis.Get()"
// that is in a header to avoid it -- let's do this later!
func init() {

	initDefaultDirs()

	configFile := filepath.Join(BrooceConfigDir, "brooce.conf")
	bytes, err := ioutil.ReadFile(configFile)
	if err != nil {
		log.Println("Unable to read config file", configFile, "so using defaults!")
	} else {
		err = json.Unmarshal(bytes, &Config)
		if err != nil {
			log.Fatalln("Your config file", configFile, "seem to have invalid json! Please fix it or delete the file!")
		}
	}

	if BrooceConfigDir != "" {
		log.Println("ConfigDir:", BrooceConfigDir)
	}
	if BrooceLogDir != "" {
		log.Println("LogDir:", BrooceLogDir)
	}

	initDefaultJobOptions()
	initDefaultConfig()

	if !util.FileExists(configFile) {
		if bytes, err := json.MarshalIndent(&Config, "", "  "); err == nil {
			err = ioutil.WriteFile(configFile, bytes, 0644)
			if err != nil {
				log.Println("Warning: Unable to write default config file to", configFile, ", error was:", err)
			} else {
				log.Println("We wrote a default config file to", configFile)
			}
		}
	}

	initThreads()
}

func initDefaultDirs() {

	if BrooceConfigDir == "" {
		BrooceConfigDir = filepath.Join(os.Getenv("HOME"), ".brooce")
	}

	if !util.IsDir(BrooceConfigDir) {
		err := os.Mkdir(BrooceConfigDir, 0755)
		if err != nil {
			log.Fatalln("Unable to create directory", BrooceConfigDir, ":", err)
		}
	}

	if BrooceLogDir == "" {
		BrooceLogDir = filepath.Join(BrooceConfigDir, "logs")
	}

	// for backward compatibility, but you can create it instead, or remove this, if you wish
	if !util.IsDir(BrooceLogDir) {
		BrooceLogDir = BrooceConfigDir
	}

}

func initDefaultConfig() {
	if Config.ClusterName == "" {
		Config.ClusterName = "brooce"
	}

	if Config.ClusterLogName == "" {
		Config.ClusterLogName = strings.Replace(Config.ClusterName, "{", "", 1)
		Config.ClusterLogName = strings.Replace(Config.ClusterLogName, "}", "", 1)
	}

	if Config.ProcName == "" {
		Config.ProcName = fmt.Sprintf("%v-%v", myip.PublicIPv4(), os.Getpid())
	}

	if !Config.Web.Disable {
		if Config.Web.Addr == "" {
			Config.Web.Addr = ":8080"
		}

		if !Config.Web.NoAuth && (Config.Web.Username == "" || Config.Web.Password == "") {
			Config.Web.Username = "admin"
			Config.Web.Password = util.RandomString(8)
			log.Printf("You didn't specify a web username/password, so we generated these: %s/%s", Config.Web.Username, Config.Web.Password)
		}

		if Config.Web.BasePath != "" {
			Config.Web.BasePath = strings.Trim(Config.Web.BasePath, "/")
			Config.Web.BasePath = "/" + Config.Web.BasePath
		}

		Config.Web.CertFile = cleanpath(Config.Web.CertFile)
		Config.Web.KeyFile = cleanpath(Config.Web.KeyFile)
	}

	if Config.Queues == nil {
		Config.Queues = []Queue{
			{
				Name:    "common",
				Workers: 1,
			},
		}
	}

	if Config.Redis.Host == "" {
		Config.Redis.Host = "localhost"
	}

	if !strings.Contains(Config.Redis.Host, ":") {
		Config.Redis.Host = Config.Redis.Host + ":6379"
	}

	if Config.Suicide.Enable {
		if Config.Suicide.Command == "" {
			Config.Suicide.Command = "sudo shutdown -h now"
		}

		if Config.Suicide.Time <= 0 {
			Config.Suicide.Time = 600
		}
	}

	if Config.Path != "" {
		os.Setenv("PATH", os.Getenv("PATH")+string(os.PathListSeparator)+Config.Path)
	}

	if Config.GlobalJobOptions == (JobOptions{}) {
		Config.GlobalJobOptions = DefaultJobOptions
	}
}

func cleanpath(path string) string {
	if path == "" || strings.HasPrefix(path, string(os.PathSeparator)) {
		return path
	} else if strings.HasPrefix(path, "~/") {
		return filepath.Join(os.Getenv("HOME"), strings.TrimPrefix(path, "~/"))
	}

	return filepath.Join(BrooceConfigDir, path)
}
